5.0 docs: Make new multicore chapters easier to discover, and emphasize impact on C bindings

edwintorok opened the pull request:
I've almost missed the new fine chapters in the manual about the memory model and parallelism, since it wasn't mentioned in the usual places I look for newly added content (language extensions, and .mli docs of new modules).
Looking a bit more closely I did find them, but considering that this is quite a prominent feature for 5.0, add a few more links to make it easier for people to discover this documentation.

Also explicitly call out the difference between the previous global runtime lock in C bindings and the domain lock. Currently this is only implied by the wording "domain lock", and IIUC this makes some C bindings that are currently working correctly have some subtle bugs in 5.0 if they are used from multiple domains, so might be worthwhile explicitly pointing this out to make bugs like this easier to catch. I hope I got this right, please correct me if I'm wrong.
(this is not just a theoretical worry, I've been recently auditing some C bindings for no-naked-pointer-safety, and came across a comment that initialized a 'static' local variable in a function with a comment saying this is safe because of the OCaml runtime lock, well not anymore...)

Draft PR:
* should I send these changes to master first?
* what do you think of the  'language extension' section addition? Should it be just a link to the other 3 pieces of docs?
(it is not strictly speaking an extension of the language *syntax*, but it is an extension of the language's semantics, so having an entry in the "language changelog" chapter of the manual seems appropriate) However I don't want to duplicate any information present in the tutorials already, so I tried to add just some very minimal context to each link.

edwintorok committed:
Manual: add note about the change in behaviour of the runtime lock with multi-domain

It is not longer safe to assume that only one C bindings runs at a time,
even if that C binding never releases the OCaml runtime lock:
the OCaml program could have multiple domains, and the lock is now a
"domain lock".

Previously C bindings may have assumed that it is safe to
manipulate/initialize C global state as long as they do not release the
OCaml runtime lock, but that is no longer the case.
E.g. a C binding having a local `static` value is no longer safe, and
worth explicitly calling out in the manual!

Document in the parallelism section of the manual, and add a link to it from
the 'Interfacing with C chapter'.

Suggested-by: KC Sivaramakrishnan <kc@kcsrk.info>
Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok committed:
Manual: make new chapters on memory model and parallelism easier to discover

When I look at learning what is new in a release of OCaml I look at:
* Language extensions (an approx chronological order of language "diffs")
* The documentation in the .mli of newly introduced modules
* The Changes entry (which did contain a reference, albeit buried deep
  among other entries)

In this case the memory model and parallelism got their own chapters in
the main manual (which is well deserved), but it is also easy to miss
(it is not in the usual places I used to look for new information).

Improve this:

* mention the manual entry in the Atomic and Domain .mli docs.

* mention Effect handler manual entry in Effect.mli

* mention the new manual chapters in the top-level multicore changes
  entry where people are most likely to notice

The mention in the .mli refers to the chapter title, there isn't a convenient
and reliable way to add a link to html with ocamldoc/odoc at the moment
({%latex:\ref%} could be used for PDF).
This can be revisited in the future e.g. after some tooling changes.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

kayceesrk commented on manual/src/cmds/intf-c.etex:
```diff
@@ -2434,6 +2434,14 @@ is held, therefore the C code has full access to the facilities of the
 run-time system.  However, no other thread in the same domain can execute
 OCaml code concurrently with the C code of the primitive.
 
+This implies that C code running in other domains *can* run in parallel
```
How about 

> During parallel execution with multiple domains, C code running on a domain may run in parallel with any C code running in other domains even if neither of them has released the ``domain lock''. Prior to OCaml 5.0, C bindings may have assumed that if the OCaml runtime lock is not released, then it would be safe to manipulate global C state (e.g. initialise a function-local `static` value). This is no longer true in the presence of parallel execution with multiple domains.

edwintorok replied:
Thanks, I added this text to the parallelism manual, and linked to it from the C manual (the parallelism manual is where it goes in more detail in how things work, so it seemed more appropriate to have the details there).

kayceesrk commented on manual/src/refman/exten.etex:
```diff
@@ -123,5 +123,30 @@ that are implemented in OCaml, but not described in chapter \ref{c:refman}.
+following semantic changes take effect:
+\begin{description}
+
+  \item[DRF-SC guarantee] is a consistent semantics for data race free parts of the program
```
DRF-SC holds even for sequential programs trivially since there are no data races. I wouldn't classify this as a semantic change that takes effect when the OCaml program has more than one domain.

kayceesrk commented on manual/src/refman/exten.etex:
```diff
@@ -123,5 +123,30 @@ that are implemented in OCaml, but not described in chapter \ref{c:refman}.
 %HEVEA\cutname{effects.html}
 \input{effects.tex}
 
+\section{s:memory-model}{Memory model for parallel OCaml programs}
```
It doesn't seem appropriate to add the memory model in the language extension section as it is not.

edwintorok replied:
I've dropped the new section, instead just added a link next to refman link: language extensions section says it documents things not present in refman, and parallelism is neither in refman nor in language extensions.

kayceesrk replied:
Parallelism and the memory model are in the tutorial section

https://github.com/ocaml/ocaml/blob/75b8d68064472a160aa4ee0eb6029561615c6a80/manual/src/allfiles.etex#L52-L53

![image](https://user-images.githubusercontent.com/410484/207564454-cbb38cd9-19f1-47da-8867-3d19b61263b4.png)

Octachron replied:
I agree with @kayceesrk that parallelism is not in its place in the extension chapter, since it is no more a language construct than threads were in OCaml 4.

edwintorok replied:
ok I'll remove it

edwintorok replied:
```
git diff origin/5.0 --stat
 manual/src/cmds/intf-c.etex           |  1 +
 manual/src/tutorials/parallelism.etex | 10 ++++++++++
 otherlibs/systhreads/thread.mli       |  7 +++++--
 stdlib/atomic.mli                     |  2 ++
 stdlib/domain.mli                     |  6 ++++++
 stdlib/effect.mli                     |  2 ++
 6 files changed, 26 insertions(+), 2 deletions(-)
```

kayceesrk commented on manual/src/refman/exten.etex:
```diff
@@ -123,5 +123,30 @@ that are implemented in OCaml, but not described in chapter \ref{c:refman}.
+%HEVEA\cutname{lang_memorymodel.html}
+(Introduced in 5.0)
+
+OCaml 5.0 has introduced support for running parallel OCaml code in the same
```
Is this text useful? For the reader to see this text, they should already be in the language extension section.

edwintorok replied:
dropped

kayceesrk commented on stdlib/atomic.mli:
```diff
@@ -17,6 +17,8 @@
 
 (** Atomic references.
 
+    See '10 Memory model: The hard bits' in the manual.
```
This is very useful. 

Optional: The absolute reference to the manual chapter will need to be kept in sync with any manual changes. Is there a way to do it better.

kayceesrk replied:
We should add a similar reference for `stdlib/effect.mli`.

edwintorok replied:
I was trying to look for existing places that would have such references, but either I was in looking in the wrong place or there aren't any. For now I'll drop the number and just refer to the title, that should still be easy enough for the user to find (well ideally this would be a clickable link, maybe I need to emit some LaTex for that from ocamldoc I'll try to take a look how to do that)

edwintorok replied:
Ok, I've added this for the PDF and HTML manuals by manually emitting a '\ref' and 'a href' and pushed a commit to show how this can be done for the 'Atomic' module, and the chapter title in the generated module doc is now clickable.

I'm not sure this is the right approach (first class support for this in ocamldoc/odoc might be better), but in the short term would you like me to add these manual links to the other modules too?

Octachron replied:
Unfortunately, it is only possible to add references to the manual in the API  documentation in the PDF output. For the html, the issue is that the API documentation and manual html output are generated independently. It is still safe to link from the manual to the API documentation because the standard library has some backward compatibility constraint which means that we can be reasonably be certain than an link to `Stdlib.List` can be given a canonical address. This is not clearly the case for the manual. Consequently, I think that citing the manual section name is the best short term solution .

edwintorok replied:
I've removed the %latex/%html and added a note in the commit message explaining the limitation and that this can be revisited in the future.

edwintorok committed:
Manual(Thread): mention domains

Mention that concurrency only applies within a domain (if there are other
domains this thread may run in parallel with those).
And add a link to Domain.spawn if the user is interested in parallelism instead
of concurrency.
(The Domain module's documentation contains a link to the manual chapter that
explains the distinction)

In other languages threads are used for parallelism too, so it is worth
pointing out that another mechanisms exists for that.

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

Octachron commented on otherlibs/systhreads/thread.mli:
```diff
@@ -23,7 +23,8 @@ type t
    in which the function application [funct arg]
-   is executed concurrently with the other threads of the program.
+   is executed concurrently with the other threads of the program in the same
+   domain.
```
I would simplify the sentence to:
```
is executed concurrently with the other threads of the domain.
```

Octachron commented on stdlib/atomic.mli:
```diff
@@ -17,6 +17,8 @@
 
 (** Atomic references.
 
+    See {%html:<a href="../parallelism.html">%}'Memory model: The hard bits' chapter{%latex:~\ref{c:memorymodel}%}{%html:</a>%} in the manual.
```
This html link seems far too brittle to me. Typically, it would require rewriting when generating the manual for `ocaml.org`.

Octachron replied:
See Octachron/ocaml@15db5a636297e4939e5926d38eaa2fee09bf5231 for the needed postprocessing if you prefer to go in this direction.

edwintorok replied:
Probably best to remove both the latex and html link for now and revisit this in the future

edwintorok commented:
Thanks for the comments, before I mark the PR as 'ready for review' should I add a Changes entry for this in the manual section, or should I add a no changes entry needed instead? (I'd go with the latter, there are plenty of changes entries for multicore and ocaml 5.0 related docs already, this is just a tiny enhancement).

[also as a general point when responding to review requests I usually mark conversations as resolved on github when I think I've addressed the reviewer's comment, but if you prefer I can also leave them open for the reviewer to close when they're happy, and just mark it with a thumbs up or comment that I've done it, which is the preferred way?]

kayceesrk commented:
The PR makes useful additions. I’m ok to merge this. Reg Changes entry, I’ll leave it for @Octachron to answer.

Octachron commented:
@edwintorok if you don't mind, a change entry (in the 5.0 section) would be nice. The change is an improvement to the documentation.

edwintorok committed:
Manual: add changes entry for manual improvements

Signed-off-by: Edwin Török <edwin.torok@cloud.com>

edwintorok commented:
Added changes entry (and tweaked title of this PR to match).
